File: cec2021/R/cec2021.R
#' @export

cec2021 = function(i, x, suite) {
  if (base::missing(i)) base::stop("Missing argument; 'i' has to be provided !")

  if (base::missing(x)) base::stop("Missing argument; 'x' has to be provided !")

  if (base::is.numeric(i) && i >= 1 && i <= 30) {
    if (base::is.vector(x)) {
      row = 1
      col = base::length(x)
    } else if (base::is.matrix(x)) {
      row = base::nrow(x)
      col = base::ncol(x)
    } else {
      base::stop("x should be a vector or a matrix")
    } 

    if (!(col %in% c(2, 10, 20))) {
      base::stop("Invalid argument: Only 2, 10, and 20  dimensions/variables are allowed !")
    }
    extdatadir = base::system.file("extdata", package = "cec2021")
    f = base::.C("cec2021",
      extdatadir = as.character(extdatadir),
      i = base::as.integer(i), x = base::as.double(x), row = base::as.integer(row),
      col = base::as.integer(col), f = double(row), suite = base::as.character(suite),
      PACKAGE = "cec2021"
    )$f
  } else {
    base::stop("Invalid argument: 'i' should be an integer between 1 and 30 !")
  }
  return(f)
}

File: cec2021/src/cec2021.c
#include <stdio.h>
#include <stdlib.h>
#include <R.h>

#include "interface.h"

double *OShift, *M, *y, *z, *x_bound;
int ini_flag = 0, n_flag, func_flag, *SS;
char *extdata;

void cec2021(char **extdatadir, int *i, double *X, int *row, int *col,
             double *f, char *suite) {
  int r, c;
  double *x;

  extdata = *extdatadir;

  x = (double *)malloc(*col * sizeof(double));

  for (r = 0; r < *row; r++) {
     R_CheckUserInterrupt();

    for (c = 0; c < *col; c++) {
      x[c] = X[r + *row * c];
    }
    cec21_func(x, &f[r], *col, 1, *i, suite);
  }

  free(x);
}
